From 5c37253d1fefd86cae62fe1bad4b5a363b1bf101 Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Wed, 5 Nov 2025 12:13:20 -0500
Subject: [PATCH] tls: adjust recv return with async crypto and failed copy to
 userspace

jira VULN-136507
cve-pre CVE-2025-39682
commit-author Sabrina Dubroca <sd@queasysnail.net>
commit 85eef9a41d019b59be7bc91793f26251909c0710

process_rx_list may not copy as many bytes as we want to the userspace
buffer, for example in case we hit an EFAULT during the copy. If this
happens, we should only count the bytes that were actually copied,
which may be 0.

Subtracting async_copy_bytes is correct in both peek and !peek cases,
because decrypted == async_copy_bytes + peeked for the peek case: peek
is always !ZC, and we can go through either the sync or async path. In
the async case, we add chunk to both decrypted and
async_copy_bytes. In the sync case, we add chunk to both decrypted and
peeked. I missed that in commit 6caaf104423d ("tls: fix peeking with
sync+async decryption").

Fixes: 4d42cd6bc2ac ("tls: rx: fix return value for async crypto")
	Signed-off-by: Sabrina Dubroca <sd@queasysnail.net>
	Reviewed-by: Simon Horman <horms@kernel.org>
Link: https://lore.kernel.org/r/1b5a1eaab3c088a9dd5d9f1059ceecd7afe888d1.1711120964.git.sd@queasysnail.net
	Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit 85eef9a41d019b59be7bc91793f26251909c0710)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 net/tls/tls_sw.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/tls/tls_sw.c b/net/tls/tls_sw.c
index 6b8e9e69e733..599a0f96c571 100644
--- a/net/tls/tls_sw.c
+++ b/net/tls/tls_sw.c
@@ -2117,6 +2117,9 @@ int tls_sw_recvmsg(struct sock *sk,
 		else
 			err = process_rx_list(ctx, msg, &control, 0,
 					      async_copy_bytes, is_peek, NULL);
+
+		/* we could have copied less than we wanted, and possibly nothing */
+		decrypted += max(err, 0) - async_copy_bytes;
 	}
 
 	copied += decrypted;
-- 
2.47.3

